File: data/pop.js
const mysql = require('mysql');
const fs = require('fs');
const path = require('path');

const colors_used = path.join(__dirname, 'data', 'The Joy Of Painting - Colors Used.csv');
const episode_dates = path.join(__dirname, 'data', 'The Joy Of Painting - Subject Matter.txt');
const subject_matter = path.join(__dirname, 'data', 'The Joy Of Painting - Episode Dates.txt');

const connection = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'root',
  database: 'joy_of_painting',
  multipleStatements: true
});

// Read colors used for episodes table
function colorsUsedforEpisodes() {
  return new Promise((resolve, reject) => {
    fs.readFile(colors_used, 'utf8', (err, data) => {
      if (err) {
        console.error('Error reading colors used file:', err);
        reject(err);
        return;
      }
      // Parse the CSV data and extract rows
      const rows = data.trim().split('\n').map(row => row.split(','));
      let episodes = [];
      for (let i = 1; i < rows.length; i++) {
        episodes.push([rows[i][3], rows[i][4], rows[i][5], rows[i][2], rows[i][7]]);
      }
      resolve(episodes);
    });
  });
};

function datesForEpisodes() {
  return new Promise((resolve, reject) => {
    fs.readFile(episode_dates, 'utf8', (err, data) => {
      if (err) {
        console.error('Error reading episode dates file:', err);
        reject(err);
        return;
      }
      // Parse the text data and extract rows
      const rows = data.trim().split('\n').map(row => row.split('('));
      let dates = [];
      for (let i = 0; i < rows.length; i++) {
        let item_date = new Date(rows[i][1].split(')')[0]);
        item_date = item_date.toLocaleDateString();
        dates.push(item_date.toString());
      }
      resolve(dates);
    });
  });
}

function mergeData(data1, data2) {
  for (let i = 0; i < data1.length; i++) {
    data1[i].push(data2[i]);
  }
  return data1;
}

// Call the functions to read and merge data from files
Promise.all([colorsUsedforEpisodes(), datesForEpisodes()])
  .then(([colors_data, dates_data]) => {
    const mergedData = mergeData(colors_data, dates_data);
    console.log(mergedData[mergedData.length - 3]);
    const sql = "INSERT INTO episodes (title, season_number, episode_number, painting_img_src, painting_yt_src, air_date) VALUES ?";

    connection.query(sql, [mergedData], (err, results, fields) => {
      if (err) {
        console.error('Error inserting data into episodes table:', err);
        return;
      }
      console.log('Data inserted successfully into episodes table.');
    });
  })
  .catch((err) => {
    console.error('Error:', err);
  })
  .finally(() => {
    // Close the database connection
    connection.end((endErr) => {
      if (endErr) {
        console.error('Error closing MySQL connection:', endErr);
      } else {
        console.log('MySQL connection closed.');
      }
    });
  });

File: pop.py
import pymysql
import pandas as pd
import os

# Configuration for the MySQL connection
db_config = {
    'host': 'localhost',
    'user': 'root',
    'password': 'root',
    'database': 'joy_of_painting'
}

# File paths
colors_used = './data/The Joy Of Painting - Colors Used'
episode_dates = './data/The Joy Of Painting - Subject Matter'

# Create a MySQL connection
try:
    connection = pymysql.connect(**db_config)
except pymysql.MySQLError as e:
    print(f"Error connecting to MySQL: {e}")
    raise

def read_csv_file(file_path, columns):
    if os.path.exists(file_path):
        df = pd.read_csv(file_path)
        print(f"Columns in {file_path}: {df.columns.tolist()}")  # Print columns for debugging
        if all(col in df.columns for col in columns):
            return df[columns].values.tolist()
        else:
            raise ValueError(f"None of the columns {columns} are in the DataFrame")
    else:
        raise FileNotFoundError(f"File {file_path} does not exist")

def colors_used_for_episodes():
    columns = ['painting_title', 'season', 'episode', 'img_src', 'youtube_src']
    return read_csv_file(colors_used, columns)

def dates_for_episodes():
    columns = ['EPISODE']  # Update to actual column name in the CSV file
    data = read_csv_file(episode_dates, columns)
    # Convert to list of strings if necessary
    return [str(episode[0]) for episode in data]

def merge_data(data1, data2):
    # Use None as a placeholder for missing air_date
    return [row1 + [None] for row1, row2 in zip(data1, data2)]

try:
    colors_data = colors_used_for_episodes()
    dates_data = dates_for_episodes()
    merged_data = merge_data(colors_data, dates_data)
    print(merged_data[:3])  # Print first few rows for debugging
    
    # Alter the table to allow NULL values for air_date
    alter_table_sql = "ALTER TABLE episodes MODIFY air_date DATE NULL"
    with connection.cursor() as cursor:
        cursor.execute(alter_table_sql)
        connection.commit()

    # Insert data into the database
    sql = """
    INSERT INTO episodes (title, season_number, episode_number, painting_img_src, painting_yt_src, air_date)
    VALUES (%s, %s, %s, %s, %s, %s)
    """
    with connection.cursor() as cursor:
        for row in merged_data:
            cursor.execute(sql, tuple(row))
        connection.commit()
        print('Data inserted successfully into episodes table.')

except Exception as e:
    print(f"Error: {e}")

finally:
    connection.close()
    print('MySQL connection closed.')
